### ISSUE TITLE
Structured strings template and camel cased Localizable.strings

### ISSUE THREAD

**Issue by alessandroorru** [SwiftGen/SwiftGen#257]
My Localizable.strings have a bunch of keys similar to this one, all correctly camelCased. Some keys also includes uppercased acronyms.

`"myViewController.userABC.nameText" = "My localization";`

The problem is that using the dot-syntax-swift3 template I get something similar to this:

```
enum Myviewcontroller {
    enum Userabc {
        static let nameText = ...
    }
}
```

As you can see all the enum names lose the camel case of the original Localizable key. 

I think that the right behavior should be to preserve the camel case, just uppercasing the first letter of the enum's name.

The change is simple (I already did that and can make a PR), but it will break a bunch of tests because it seems that this behavior has been done by purpose to merge together keys with same name but different capitalization.

What is your opinion on this? I already know that I could use snake case, but I don't like that and I will lose anyway the acronyms capitalization.

**Comment by AliSoftware** [comment]
Hello

Which version of SwiftGen are you using?
The latest 4.1 [fixed a bunch of issues](https://github.com/AliSoftware/SwiftGen/blob/master/CHANGELOG.md) both with the `dot-syntax-swift3` template and with the canal case filter used with that template, so I'd you're not on the latest version yet this might already been fixed and you would just need to update 😉

Also could you confirm which template you're using? You mention the `structured` template in the title of the issue but the `dot-syntax-swift3` in the description so just wanna be sure

Finally, did you try to customize the template to your needs? It's really easy to do, and after all that's what templates are for too, so people can adapt to their needs when the bundled templates don't fit your exact needs (see `documentation/Templates.md` for more info)

**Comment by alessandroorru** [comment]
@AliSoftware Yeah I'm using the 4.1 version and the dot-syntax-swift3. 

Anyway this isn't a template problem: I checked out the repository and debugged it. 

The problem is here on [Contexts.swift](https://github.com/AliSoftware/SwiftGen/blob/master/GenumKit/Stencil/Contexts.swift) lines 287-295:

```
    // make key paths unique
    let uniqueNextLevelKeyPaths = Array(Set(
      nextLevelKeyPaths.map { keyPath in
        keyPath.map({
          $0.capitalized.replacingOccurrences(of: "-", with: "_")
        }).joined(separator: ".")
      }))
      .sorted()
      .map { $0.components(separatedBy: ".") }
```

the `capitalized` version of a string like "camelCase" will return "Camelcase" and not "CamelCase". So the template will already receive a broken string.

**Comment by djbe** [comment]
Hmmm, that's weird though, because:

**Comment by alessandroorru** [comment]
@djbe Ops I'm sorry, I correct myself: only the enum names get messed up, not the key tail. I update my main post.

Try a simple test: add "ObjectOwnership.anotherString" to the Localizable.strings. Now the "ObjectOwnership" will need to be converted to a subenum (since it has a subkey) and it will result in "enum Objectownership" using the `dot-syntax-swift3`.

This is also easily understandable by looking at the current Localizable.strings:

```
"settings.navigation-bar.title.even.deeper" = "Settings";
"SETTINGS.USER_PROFILE_SECTION.HEADER_TITLE" = "User Profile Settings";
"seTTings.uSer-proFile-sectioN.footer_text" = "Here you can change some user profile settings.";
"settings.navigation-bar.title.even.deeper.than.we.can.handle" = "DeepSettings";
```

If you use `dot-syntax-swift3` you will have a single `enum Settings` instead of `Settings`, `SETTINGS` and `SeTTings`. That's why I asked if this has been done by purpose: it seems you preferred to merge together wrongly cased strings in a single enum instead of having them preserve the case. Personally, I think it is a user fault if the Localizable.strings is messed up, and it should be corrected there, not by Swiftgen.

**Comment by djbe** [comment]
Still, once those get converted to a correct `swift` naming, for all cases you'd get `Settings`.

Now your case is special in the sense that it's a key/path that's camel-cased instead of snake-cased. #233 added logic for this that checks the contents of a string before applying a transformation.

Maybe the solution could be to simply modify `$0.capitalized` to `$0.snakecased.capitalized`. (we don't have those string extensions, but you get the idea, just use the filters we have)

**Comment by alessandroorru** [comment]
How would this solve? I think snake casing it would would result in `enum Object_ownership`. Am I wrong?

I think the correct one would be `titlecase`. I quickly tried converting `$0.capitalized` to `(try! StringFilters.titlecase($0) as! String)` and it would work.

**Comment by AliSoftware** [comment]
I think what @djbe meant is that `"ObjectOwnership".shakecased` would indeed result in `"Object_ownership"` but applying `.camelcased` to that result will return `"ObjectOwnership"`, so `"ObjectOwnership".snakecased.camelcased` would result in the proper `enum ObjectOwnership` name in the end (indeed in that case the end result is the same as the input result, but we'd still need to apply this filter to keep other cases too continue to work the same as now)

**Comment by AliSoftware** [comment]
Also, keeping the key unchanged causes issues for other people, like people who use the naming conventions like `SETTINGS_FOO` for their `Localizable.strings` keys, but we still want the generated enum to follow the Swift 3 guidelines. So we don't want to generate `enum SETTINGS` but `enum Settings` to be Swift3-compliant.

And we don't want to impose the users to format their `Localizable.strings` keys in a specific way, because they could have other reasons to make them all uppercased, like maybe it's historic in their project and they shouldn't have to change all their keys themselves, especially given that if we use the key name parts unchanged they would need to be sure to capitalize each part… except the last (which generates a `case` and not a nested `enum`)… so that would force them to name their keys… based on which template they chose and based on constraints imposed by the template, even if the template used might not be the only thing imposing a constraint (maybe the people translating the app already use all-caps keys or use a translation tool that requires that, etc)

So, basically, that's why we apply the string-casing filters ourselves in the template, to have homogeneous behaviour whatever the constraints and standards user use in their `Localizable.strings`. We don't really do it "in order to group `SETTINGS`, `Settings` and `SeTTinGs` together in a single enum on purpose", it's more that we adapt the key case according to Swift 2 or Swift 3 naming conventions, and if it happens that multiple keys with same text but different case ends up resulting to the same cased string after applying the case filters, they end up in the same `enum` to be consistent and to make it compile.

**Comment by alessandroorru** [comment]
@AliSoftware this is absolutely ok for me. The problem is that you are talking about "string-casing filters", but in this case you lose the case **before** the templates receive the keys. The template receives `Objectownership` instead of `ObjectOwnership`, so there is no filter that could make it work unless we fix the code.

**Comment by AliSoftware** [comment]
Indeed, you're right.
But I'm not sure how we could handle that. I mean, if we don't filter the key name by code but let the template filter it (which would make more sense), then there's the problem of keys that would generate the same output _after_ the filter has been applied by the template. This means that if we keep the keys unchanged in the code, the Stencil context could look like this:

```json
{
  "settings": ["foo", "bar"],
  "SeTTinGs": ["baz"]
}
```

And then the template won't easily be able to merge all the keys which, once case-filtered, will look the same, so when it will iterate over each key to generate an enum for each, it will generate one enum for each key even if they end up having the same name once the key is case-filtered by the template:

```swift
enum Settings {
  case foo
  case bar
}
enum Settings {
  case baz
}
```

---

Maybe one solution would be to make the key all-lowercase for indexing (and that's what the template will iterate onto), but in the value associated to the key, repeat the original key unchanged. Like:

```json
{
  "settings": { "key": "settings", strings: [
    { "key": "settings_foo", "value": "foo" },
    { "key": "settings_bar", "value": "bar" },
    { "key": "SeTTinGs_baz", "value": "baz" }
 ]}
}
```

But again, that assumes that the filters applied by the template only change the case. What if another template decides to apply a different filter, say an hypothetic `removeNonAlpha` filter to only keep `[A-Za-z]` and remove numbers and special chars from the generated enum name, then keys `"sett0ing1"` and `"SeT2TinG3s"` will generate the `"sett0ing1"` and `"set2ting3s"` lowercase keys for indexing and thus will end up generating 2 `enums`, which will end up both being named `enum Settings` anyway once the filter `removeNonAlpha` is applied… So that doesn't even solve the problem either.

So perfect solution would be to be able to "merge together all the keys in the Context that will end up generating the same enum name as output"… all via template code, so that it can merge them depending on what filters would be used to generate the `enum` name (filters which could be different from template to template) — and not merge them in the SwiftGen code ahead of time guessing what filter could be applied. But not sure if that's actually feasible, and if so, what would the Stencil code look like (and will it be readable enough?!)

**Comment by alessandroorru** [comment]
Yeah now I see the point and I partially agree with you. I think there is no silver bullet.

Anyway, you said: "we don't want to impose the users to format their Localizable.strings keys in a specific way". I agree with this, but in my case you are actually forcing me to convert my Localizable.strings to snake case: I'm pretty sure I'm not the only one in the world with camelCased Localizable.strings, and this creates a really, really ugly and hard to read code. 

On the other hand, what I believe anyway is that if a user have two keys, one with "settings" and the other with "seTTings", it would be acceptable if Swiftgen couldn't generate a valid file: the user actually messed up the Localizable.strings mixing up name formats, and he would need to just **fix** it, not to change the whole file format.

I also add that if you create a `removeNonAlpha` with the current version of Swiftgen you would have the same problem: the generated template would end up in a duplicated `enum Settings`. There is no solution for this.

**Comment by alessandroorru** [comment]
Meanwhile I converted camelCase to snake_case. I also created a [script](https://gist.github.com/alessandroorru/59811fde7f6c995ea41ae892cd3fc1ff) that should convert the occurrences of NSLocalizedString into dot-syntax-swift3 compatible calls, based on the original Lutzifer script.

**Comment by ivanbruel** [comment]
Ended up on this issue for exactly the same reasons, maybe having the developer make sure the strings are uniquely set would be a better way, even if its just a parameter in the strings generation executable. It's troublesome that we can't work around this without converting everything to snake case.

**Comment by AliSoftware** [comment]
I agree and actually the more I think about it and talk about it with others the more I think people are not interested in respecting the Swift 3 naming conventions for those specific cases, because even though you want to respect the conventions for your classes and types in your application, in the case of constants you don't really care in practice that those constants are enums, and you'd prefer using `MyViewController.userABC.nameText` (which would the the exact case used in your `Localizable.strings`, so it even means that you can copy/paste the key name from there) to access that constant, even if that means that there would be an `enum userABC` not respecting the Swift naming and casing conventions there, because (1) you don't really case that it's an enum when you call it and (2) if you care about it and want to respect the Swift naming conventions anyway, it's still _your_ job do to so (so your job to adapt your `Localizable.strings` keys manually)

So I think the end solution would be to stop modifying the case in the parser code of SwiftGenKit as you suggested… but I'm even wondering if it wouldn't make sense to ship a template that doesn't change the case of the enum names either. At least propose that as an alternate template.

**Comment by alessandroorru** [comment]
@AliSoftware I agree with your point of view. I think that the really, really important thing is that the strings parser doesn't manipulate the strings at all, and treat them as they are. The optional conversion into a specific format should be handled only on the template side, so that each developer can eventually create a template that conforms to their very specific needs.

**Comment by dchohfi** [comment]
Hey guys, any thoughts on this?

**Comment by AliSoftware** [comment]
For ref, this is being currently addressed in PR #398
